Mathlib/NumberTheory/ModularForms/NormTrace.lean

```diff
@@ -0,0 +1,176 @@
+  bdd_at_cusps' h γ := by
+    rintro rfl
+    simp_rw [SlashInvariantForm.norm, IsBoundedAtImInfty, Filter.BoundedAtFilter]
+    let := Fintype.ofFinite 𝒬
+    rw [Nat.card_eq_fintype_card, ModularForm.prod_fintype_slash]
+    apply Asymptotics.IsBigO.const_smul_left
+    rw [show (1 : ℍ → ℝ) = (fun x ↦ ∏ (i : 𝒬), 1) by ext; simp, Finset.prod_fn]
+    refine .finsetProd (Quotient.forall.mpr fun ⟨r, hr⟩ _ ↦ (translate f _).bdd_at_cusps' ?_ γ rfl)
+    simpa using h.of_isFiniteRelIndex_conj hr
+
+variable {f} in
+lemma ModularForm.norm_ne_zero [ℋ.HasDetPlusMinusOne] [ModularFormClass F 𝒢 k]
+    (hf : (f : ℍ → ℂ) ≠ 0) : ModularForm.norm ℋ f ≠ 0 := by
+  contrapose! hf
+  rw [← DFunLike.coe_injective.eq_iff, coe_norm, coe_zero, prod_eq_zero_iff] at hf
+  · simpa [QuotientGroup.exists_mk] using hf
+  · exact Quotient.forall.mpr fun r _ ↦ (translate f r.val⁻¹).holo'
+
+lemma ModularForm.norm_eq_zero_iff [ℋ.HasDetPlusMinusOne] [ModularFormClass F 𝒢 k] :
+    ModularForm.norm ℋ f = 0 ↔ (f : ℍ → ℂ) = 0 := by
+  refine ⟨fun hn ↦ ?_, fun hf ↦ ?_⟩
+  · contrapose! hn
+    exact norm_ne_zero ℋ hn
+  · ext τ
+    simpa [Finset.prod_eq_zero_iff, QuotientGroup.exists_mk]
+      using ⟨1, by simpa using congr_fun hf τ⟩
+
+open scoped MatrixGroups
+
+lemma ModularForm.isZero_of_neg_weight [𝒢.IsArithmetic]
+    {k : ℤ} (hk : k < 0) (f : ModularForm 𝒢 k) : f = 0 := by
+  suffices ModularForm.norm 𝒮ℒ f = 0 by simp_all [ModularForm.norm_eq_zero_iff]
+  ext
+  -- some friction here because `levelOne_neg_weight_eq_zero` uses `Γ(1)` for the level
+  rw [@ModularFormClass.levelOne_neg_weight_eq_zero (f := ModularForm.norm 𝒮ℒ f) _ _ _]
+  · rfl
+  · rw [CongruenceSubgroup.Gamma_one_top, ← MonoidHom.range_eq_map]
+    infer_instance
+  · exact mul_neg_of_neg_of_pos hk <| mod_cast Nat.pos_of_ne_zero 𝒢.relIndex_ne_zero
+
+private lemma ModularForm.weight_zero_const₀ [𝒢.IsArithmetic] [𝒢.HasDetOne] (f : ModularForm 𝒢 0) :
+    ∃ c, (f : ℍ → ℂ) = Function.const ℍ c := by
+  -- Consider the norm of `f - (f I)`. This must be a constant, since it's a weight 0 level 1 form.
+  obtain ⟨c, hc⟩ : ∃ c, (ModularForm.norm 𝒮ℒ (f - .const (f I)) : ℍ → ℂ) = Function.const ℍ c := by
+    convert @ModularFormClass.levelOne_weight_zero_const
+      (f := ModularForm.norm 𝒮ℒ (f - .const (f I))) _ (by infer_instance)
+      (by rw [zero_mul, CongruenceSubgroup.Gamma_one_top, MonoidHom.range_eq_map]; infer_instance)
+  -- But the constant must be 0, since `f - f I` vanishes at `I`.
+  have : ModularForm.norm 𝒮ℒ (f - .const (f I)) I = 0 := by
+    simpa [Finset.prod_eq_zero_iff, QuotientGroup.exists_mk] using ⟨1, by simp⟩
+  obtain rfl : c = 0 := by simpa [hc]
+  -- So `f - f I` has zero norm, hence it's the zero form.
+  simp only [Function.const_zero, coe_eq_zero_iff, norm_eq_zero_iff, sub_eq_zero] at hc
+  exact ⟨f I, by rw [hc, ModularForm.coe_const, Function.const_apply]⟩
+
+lemma ModularForm.weight_zero_const [𝒢.IsArithmetic] (f : ModularForm 𝒢 0) :
```

Comment:
```suggestion
lemma ModularForm.eq_const_of_weight_zero [𝒢.IsArithmetic] (f : ModularForm 𝒢 0) :
```
seems to be closer to the naming convention (similarly for the `private lemma` above).